internal/impl/aws/input_kinesis_checkpointer.go

```diff
@@ -282,17 +283,35 @@ func (k *awsKinesisCheckpointer) Claim(ctx context.Context, streamID, shardID, f
 	// This allows the victim client to update the checkpoint with the final
 	// sequence as it yields the shard.
 	if len(fromClientID) > 0 && time.Since(currentLease) < k.leaseDuration {
-		// Wait for the estimated next checkpoint time plus a grace period of
-		// one second.
 		waitFor := k.leaseDuration - time.Since(currentLease) + time.Second
-		select {
-		case <-time.After(waitFor):
-		case <-ctx.Done():
-			return "", ctx.Err()
+		timer := time.NewTimer(waitFor)
+	FOR_LOOP:
+		for {
+			checkPointed, err := k.isCheckpointed(ctx, streamID, shardID, startingSequence)
+			if err != nil {
+				return "", err
+			}
+
+			if checkPointed {
+				break
+			}
+
+			// Wait for the estimated next checkpoint time plus a grace period of
```

Comment:
nit: Could we move this comment up where we create the timer. Also, could we add another comment explaining what's happening here.

Reply:
Done